Userland/Libraries/LibWeb/HTML/FormAssociatedElement.cpp

```diff
@@ -152,61 +165,248 @@ void FormAssociatedElement::reset_form_owner()
 {
-    // 1. If this element is an input element, and selectionEnd does not apply to this element, return null.
-    // NOTE: This is done by HTMLInputElement before calling this function
+    // 1. If this element is an input element, and selectionEnd does not apply to this element, return
+    //    null.
+    auto const& html_element = form_associated_element_to_html_element();
+    if (is<HTMLInputElement>(html_element)) {
+        auto const& input_element = static_cast<HTMLInputElement const&>(html_element);
+        if (!input_element.selection_or_range_applies())
+            return {};
+    }
 
-    // 2. If there is no selection, return the code unit offset within the relevant value to the character that
-    //    immediately follows the text entry cursor.
-    if (auto cursor = form_associated_element_to_html_element().document().cursor_position())
-        return cursor->offset();
+    // 2. If there is no selection, return the code unit offset within the relevant value to the
+    //    character that immediately follows the text entry cursor.
+    if (m_selection_start == m_selection_end) {
+        if (auto cursor = form_associated_element_to_html_element().document().cursor_position())
+            return cursor->offset();
+    }
 
-    // FIXME: 3. Return the code unit offset within the relevant value to the character that immediately follows the end of
-    //           the selection.
-    return 0;
+    // 3. Return the code unit offset within the relevant value to the character that immediately
+    //    follows the end of the selection.
+    return m_selection_end;
 }
 
 // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#textFieldSelection:dom-textarea/input-selectionend-3
-WebIDL::ExceptionOr<void> FormAssociatedElement::set_selection_end(Optional<WebIDL::UnsignedLong> const&)
+WebIDL::ExceptionOr<void> FormAssociatedElement::set_selection_end(Optional<WebIDL::UnsignedLong> const& value)
+{
+    // 1. If this element is an input element, and selectionEnd does not apply to this element,
+    //    throw an "InvalidStateError" DOMException.
+    auto& html_element = form_associated_element_to_html_element();
+    if (is<HTMLInputElement>(html_element)) {
+        auto& input_element = static_cast<HTMLInputElement&>(html_element);
+        if (!input_element.selection_or_range_applies())
+            return WebIDL::InvalidStateError::create(html_element.realm(), "setSelectionEnd does not apply to this input type"_fly_string);
+    }
+
+    // 2. Set the selection range with the value of this element's selectionStart attribute, the
+    //    given value, and the value of this element's selectionDirection attribute.
+    return set_selection_range(m_selection_start, value, selection_direction());
+}
+
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#selection-direction
+Optional<String> FormAssociatedElement::selection_direction() const
```

Comment:
This seems non-Optional?

Reply:
The IDL for HTMLInputElement defines optional `.selectionStart`, `.selectionEnd` and `.selectionDirection`. The IDL for HTMLTextAreaElement defines those as non-optional. I've chosen to implement the optional ones in the base class FormAssociatedElement, and let HTMLTextAreaElement reuse those by simply invoking `.value()` on the result.

Do you know of a better way to do this? I could repeat the bindings for both HTMLInputElement and HTMLTextAreaElement, but I'd end up with the same optional values in FormAssociatedElement...

Reply:
Ah I see. In that case, isn't this missing the check that would make it actually return null?
https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#dom-textarea/input-selectiondirection

> If this element is an input element, and selectionDirection does not apply to this element, return null.

It's definitely a bit awkward, spec-wise. I don't have a suggestion that doesn't leave some awkwardness on the table.

I think I'm on board with the implementation you have here - i.e. have `FormAssociatedElement` implement the spec steps as written, and then make `HTMLTextArea` apply its valid assumptions (i.e. can just call `.value()`). It might be nice to have the IDL generator do that for us somehow, but that's a future improvement ofc.

Reply:
Ah I can't believe I didn't copy over the `.selectionDirection` spec. This also exposed an issue in the reference test, `.selectionDirection` not being `null` for `<input type="date">`.